src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs

```diff
@@ -0,0 +1,506 @@
+                        {
+                            SafeKeyHandle hKey = CapiHelper.GetKeyPairHelper(
+                                CapiHelper.CspAlgorithmType.Dss,
+                                _parameters,
+                                _keySize,
+                                SafeProvHandle);
+
+                            Debug.Assert(hKey != null);
+                            Debug.Assert(!hKey.IsInvalid);
+                            Debug.Assert(!hKey.IsClosed);
+
+                            _safeKeyHandle = hKey;
+                        }
+                    }
+                }
+
+                return _safeKeyHandle;
+            }
+
+            set
+            {
+                lock (_parameters)
+                {
+                    SafeKeyHandle current = _safeKeyHandle;
+
+                    if (ReferenceEquals(value, current))
+                    {
+                        return;
+                    }
+
+                    _safeKeyHandle = value;
+                    current?.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// CspKeyContainerInfo property
+        /// </summary>
+        public CspKeyContainerInfo CspKeyContainerInfo
+        {
+            get
+            {
+                // Desktop compat: Read the SafeKeyHandle property to force the key to load,
+                // because it might throw here.
+                SafeKeyHandle localHandle = SafeKeyHandle;
+                Debug.Assert(localHandle != null);
+
+                return new CspKeyContainerInfo(_parameters, _randomKeyContainer);
+            }
+        }
+
+        /// <summary>
+        /// KeySize property
+        /// </summary>
+        public override int KeySize
+        {
+            get
+            {
+                byte[] keySize = CapiHelper.GetKeyParameter(SafeKeyHandle, Constants.CLR_KEYLEN);
+                _keySize = (keySize[0] | (keySize[1] << 8) | (keySize[2] << 16) | (keySize[3] << 24));
+                return _keySize;
+            }
+        }
+
+        public override KeySizes[] LegalKeySizes
+        {
+            get
+            {
+                return new[] { new KeySizes(512, 1024, 64) }; // per FIPS 186-2
+            }
+        }
+
+        /// <summary>
+        /// Persisted key in CSP 
+        /// </summary>
+        public bool PersistKeyInCsp
+        {
+            get
+            {
+                return CapiHelper.GetPersistKeyInCsp(SafeProvHandle);
+            }
+            set
+            {
+                bool oldPersistKeyInCsp = this.PersistKeyInCsp;
+                if (value == oldPersistKeyInCsp)
+                {
+                    return; // Do nothing
+                }
+                CapiHelper.SetPersistKeyInCsp(SafeProvHandle, value);
+            }
+        }
+
+        /// <summary>
+        /// Gets the information of key if it is a public key
+        /// </summary>
```

Comment:
Superfluous doc comment.  (It actually only gets whether or not we believe it's a public key.  And I think it only identifies that when importing from bytes, so a named public-only key probably returns false)